weatherAplication/Modules/Search/ViewModel/SearchViewModel.swift

```diff
@@ -0,0 +1,58 @@
+import Foundation
+
+protocol SearchViewModelContract {
+    var delegate: SearchViewModelDelegate? { get set }
+    var cityInfoArray: [CityInfo] { get }
+
+    func textChanged(searchTerm: String)
+}
+
+class SearchViewModel: SearchViewModelContract {
+
+    var delegate: SearchViewModelDelegate?
+    let apiClient = APIClient.sharedInstance
+
+    var debounceTimer: Timer?
+
+    var APIRequestFlag: Bool = false
+    var cityInfoArray: [CityInfo] = []
+    let delay = 0.5
+
+    func cityChosen(cityInfo: CityInfo) {
+        delegate?.cityChosen(cityInfo: cityInfo)
+    }
+
+    func textChanged(searchTerm: String) {
+        validate(searchTerm: searchTerm)
+    }
+
+    func loadCityNames(searchTerm: String) {
+        if searchTerm == "" {
+                cityInfoArray = []
+                delegate?.reloadCityNames()
+        } else {
+            self.apiClient.searchCities(searchTerm: searchTerm, completion: { [weak self] cityInfo -> Void in
+                self?.cityInfoArray = []
+                self?.cityInfoArray = cityInfo
+                self?.delegate?.reloadCityNames()
+            }, failure: { weatherError in
+                self.delegate?.showAlert(description: weatherError.localizedDescription)
+            })
+        }
+
+    }
+
+    func validate(searchTerm: String) {
```

Comment:
tak sobie myślę że tą funkcję można inaczej nazwać, bo w sumie ona nic nie waliduje tylko to jest debounce. Myślę że możesz zrobić tego debounce'a z completion, tak żeby on był w miarę uniwersalny do użycia. I dopiero z poziomu funkcji textChanged, w implementacji completion, wołać funkcję loadCityNames